File: assignments/assignment-5/HeapSort.java
public class HeapSort {

    static void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    // sink (heapify) node i down, within a heap of the given size
    static void sink(int[] a, int size, int i) {
        int L = 2 * i + 1;
        int R = 2 * i + 2;
        int largest = i;

        if (L < size && a[L] > a[largest]) largest = L;
        if (R < size && a[R] > a[largest]) largest = R;

        if (largest != i) {
            swap(a, i, largest);
            sink(a, size, largest);
        }
    }

    // bottom-up heap construction: O(n)
    static void buildMaxHeap(int[] a, int n) {
        for (int i = n / 2 - 1; i >= 0; i--) {
            sink(a, n, i);
        }
    }

    static void heapSort(int[] a) {
        int n = a.length;
        buildMaxHeap(a, n);

        // repeatedly move max (root) to the end, shrink heap, re-sink
        for (int end = n - 1; end > 0; end--) {
            swap(a, 0, end);
            sink(a, end, 0);   // heap size is now `end`
        }
    }

    static void print(int[] a) {
        for (int x : a) System.out.print(x + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        int[] a = {5, 15, 1, 3, 2, 8, 7, 9, 10, 6};

        System.out.print("Before: ");
        print(a);

        heapSort(a);

        System.out.print("After:  ");
        print(a);
    }
}

File: dsa/heap/heapSort/heapSort.c
#include <stdio.h>
int size = 0;
void swap(int a[], int i, int j) {
    int temp = a[i];
    a[i] = a[j];
    a[j] = temp;
}
void sink(int heap[], int i) {
    int L = 2 * i + 1;
    int R = 2 * i + 2;
    int max = i;
    if (L < size && heap[L] > heap[max])
        max = L;
    if (R < size && heap[R] > heap[max])
        max = R;
    if (max != i) {
        swap(heap, i, max);
        sink(heap, max);
    }
}
void heapify(int a[], int n) {
    size = n;
    for (int i = n / 2; i >= 0; i--)
        sink(a, i);
}
void heap_sort(int a[], int n) {
    heapify(a, n);
    for (int i = n - 1; i > 0; i--) {
        swap(a, 0, i);
        size--;
        sink(a, 0);
    }
}
void print(int a[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", a[i]);
    printf("\n");
}
int main() {
    int a[] = {5, 15, 1, 3, 2, 8, 7, 9, 10, 6};
    int n = sizeof(a) / sizeof(a[0]);

    printf("Before: ");
    print(a, n);

    heap_sort(a, n);

    printf("After:  ");
    print(a, n);

    return 0;
}
